moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.kt

```diff
@@ -0,0 +1,174 @@
+
+  override fun fromJson(reader: JsonReader): T? {
+    val resultsArray = arrayOfNulls<Any>(componentBindingsArray.size)
+
+    reader.beginObject()
+    while (reader.hasNext()) {
+      val index = reader.selectName(options)
+      if (index == -1) {
+        reader.skipName()
+        reader.skipValue()
+        continue
+      }
+      resultsArray[index] = componentBindingsArray[index].adapter.fromJson(reader)
+    }
+    reader.endObject()
+
+    return try {
+      @Suppress("UNCHECKED_CAST")
+      constructor.invokeWithArguments(*resultsArray) as T
+    } catch (e: InvocationTargetException) {
+      throw e.rethrowCause()
+    } catch (e: Throwable) {
+      // Don't throw a fatal error if it's just an absent primitive.
+      for (i in componentBindingsArray.indices) {
+        if (resultsArray[i] == null && componentBindingsArray[i].accessor.type().returnType().isPrimitive) {
+          throw missingProperty(
+            propertyName = componentBindingsArray[i].componentName,
+            jsonName = componentBindingsArray[i].jsonName,
+            reader = reader
+          )
+        }
+      }
+      throw AssertionError(e)
+    }
+  }
+
+  override fun toJson(writer: JsonWriter, value: T?) {
+    writer.beginObject()
+
+    for (binding in componentBindingsArray) {
+      writer.name(binding.jsonName)
+      val componentValue = try {
+        binding.accessor.invoke(value)
+      } catch (e: InvocationTargetException) {
+        throw e.rethrowCause()
+      } catch (e: Throwable) {
+        throw AssertionError(e)
+      }
+      binding.adapter.toJson(writer, componentValue)
+    }
+
+    writer.endObject()
+  }
+
+  override fun toString() = "JsonAdapter($targetClass)"
+
+  companion object Factory : JsonAdapter.Factory {
+    override fun create(type: Type, annotations: Set<Annotation>, moshi: Moshi): JsonAdapter<*>? {
+      if (annotations.isNotEmpty()) {
+        return null
+      }
+
+      if (type !is Class<*> && type !is ParameterizedType) {
+        return null
+      }
+
+      val rawType = type.rawType
+      if (!rawType.isRecord) {
+        return null
+      }
+
+      val components = rawType.recordComponents
+      val bindings = LinkedHashMap<String, ComponentBinding<Any>>()
+      val lookup = MethodHandles.lookup()
+      val componentRawTypes = Array<Class<*>>(components.size) { i ->
+        val component = components[i]
+        val componentBinding =
+          createComponentBinding(type, rawType, moshi, lookup, component)
+        val replaced = bindings.put(componentBinding.jsonName, componentBinding)
+        if (replaced != null) {
+          throw IllegalArgumentException(
+            "Conflicting components:\n    ${replaced.componentName}\n    ${componentBinding.componentName}"
+          )
+        }
+        component.type
+      }
+
+      val constructor = try {
+        lookup.findConstructor(rawType, methodType(knownNotNull(Void::class.javaPrimitiveType), componentRawTypes))
```

Comment:
Should we save off the `knownNotNull(Void::class.javaPrimitiveType)` as a top-level property?

Reply:
Is it just a better-documented `!!`? If it's used extensively yeah can be a top-level property.

Reply:
It skips the nullcheck